flowmachine/flowmachine/core/subscriber_subset.py

```diff
@@ -0,0 +1,186 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from abc import abstractmethod
+from sqlalchemy.sql import ClauseElement, select, text, column
+from .query import Query
+
+
+class SubscriberSubsetBase(Query):
+    """
+    Base class for the different types of subscriber subsets.
+    """
+
+    @property
+    @abstractmethod
+    def is_proper_subset(self):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'is_proper_subset'"
+        )
+
+    @abstractmethod
+    def apply_subset(self, sql, *, subscriber_identifier):
+        raise NotImplementedError(
+            f"Class {self.__class__.__name__} does not implement 'apply_subset'"
+        )
+
+    def _get_query_attrs_for_dependency_graph(self, analyse=False):
```

Comment:
Yes, this is a subtle one. The only reason I had to factor this out into a helper function is because the new subset classes currently derive from `flowmachine.Query` (which I'd like to change again soon-ish but which was the easiest way of avoiding ripples through the rest of the code base). As a consequence of deriving from `Query`, the `dependency_graph()` function will try to call `.explain()` and `.is_stored` on each sub-query that contributes to the given flowmachine query. In particular, this will be called for the subset classes, too, but they don't represent full calleable queries so these don't make sense. The easiest way I could see to work around this was to factor out the code which retrieved information about this query into a helper function (`._get_query_attrs_for_dependency_graph()`) and the implement a stub for this for the subset classes which just returned a dictionary of the correct shape (but without having to call `.explain()` and `.is_stored` internally.

It's not an ideal solution, but I also don't see any harm in it because the helper function `Query._get_query_attrs_for_dependency_graph()` isn't a bad idea in any case (it separates the logic which extracts info about the query from the rest of the logic in `Query.dependency_graph()` which only applies attributes for drawing the graph).

Does this make things clearer?

Reply:
To answer your question: _"Don't we still need to be able to know whether the query has been stored, the cost, etc? Is it something we cannot derive anymore?"_

Yes, we still need to be able to do it. And we're still doing it - except that it now happens in the helper function `Query._get_query_attrs_for_dependency_graph()` rather than directly in ` Query.dependency_graph()`.

Reply:
Ok, got it. So basically you are saying that `sqlalchemy` queries do not have the `explain` and `is_stored` which would break the subset class otherwise.

My point was that the Subset `_get_query_attrs_for_dependency_graph` method does not contain information about whether it has been stored, costs, etc because they are just stubs. So I was just wondering if this is a temporary state of affairs.

Reply:
Ah, I see. Yes, you are right. And I can see that the comment doesn't really clarify this. 😂 I'll tweak it.